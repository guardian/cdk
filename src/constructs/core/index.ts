export * from "./identity";
export * from "./migrating";
export * from "./parameters";
export * from "./ssm";
export * from "./stack";
