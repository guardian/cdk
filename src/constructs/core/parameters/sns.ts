import { SSM_PARAMETER_PATHS } from "../../../constants/ssm-parameter-paths";
import { isSingletonPresentInStack } from "../../../utils/singleton";
import type { GuStack } from "../stack";
import { GuStringParameter } from "./base";

/**
 * Creates a CloudFormation parameter which references an SNS topic to send
 * alarms to. By default, the bucket name is stored in an SSM Parameter called
 * `/account/services/alarm.topic.name`.
 */
export class GuAlarmTopicParameter extends GuStringParameter {
  private static instance: GuAlarmTopicParameter | undefined;

  private constructor(scope: GuStack) {
    super(scope, "AlarmTopicName", {
      description: SSM_PARAMETER_PATHS.AlarmTopic.description,
      default: SSM_PARAMETER_PATHS.AlarmTopic.path,
      fromSSM: true,
    });
  }

  public static getInstance(stack: GuStack): GuAlarmTopicParameter {
    if (!this.instance || !isSingletonPresentInStack(stack, this.instance)) {
      this.instance = new GuAlarmTopicParameter(stack);
    }

    return this.instance;
  }
}

/**
 * Creates a CloudFormation parameter which references an SNS topic to send
 * notifications to. By default, the bucket name is stored in an SSM Parameter
 * called `/account/services/alarm.topic.name`.
 */
export class GuNotificationTopicParameter extends GuStringParameter {
  private static instance: GuNotificationTopicParameter | undefined;

  private constructor(scope: GuStack) {
    super(scope, "NotificationTopicName", {
      description: SSM_PARAMETER_PATHS.NotificationTopic.description,
      default: SSM_PARAMETER_PATHS.NotificationTopic.path,
      fromSSM: true,
    });
  }

  public static getInstance(stack: GuStack): GuNotificationTopicParameter {
    if (!this.instance || !isSingletonPresentInStack(stack, this.instance)) {
      this.instance = new GuNotificationTopicParameter(stack);
    }

    return this.instance;
  }
}
