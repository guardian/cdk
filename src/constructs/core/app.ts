import { Tags } from "aws-cdk-lib";
import { Construct } from "constructs";
import type { Identity } from "./identity";
import type { GuStack } from "./stack";

export class GuApp extends Construct implements Identity {
  public readonly stack: string;
  public readonly stage: string;
  public readonly app: string;

  /**
   * A sanitised version of `app` suitable to use for the logical ID of CloudFormation resources, which must be alphanumeric.
   * Typically, only needed when wanting to create a static (deterministic) logical ID during synthesis.
   *
   * @see https://stackoverflow.com/a/20864946
   */
  public readonly appForLogicalId: string;

  public readonly parent: GuStack;

  // eslint-disable-next-line custom-rules/valid-constructors -- `scope` cannot be `GuApp` as that's recursive
  constructor(scope: GuStack, id: string) {
    super(scope, id);

    this.stack = scope.stack;
    this.stage = scope.stage;
    this.app = id;

    const titleCaseApp = id.charAt(0).toUpperCase() + id.slice(1);
    this.appForLogicalId = titleCaseApp.replace(/[\W_]+/g, "");

    this.parent = scope;

    Tags.of(this).add("App", id);
  }
}
