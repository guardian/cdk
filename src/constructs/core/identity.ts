import type { IConstruct } from "@aws-cdk/core";
import { Tags } from "@aws-cdk/core";

export interface StackStageIdentity {
  stack: string;
  stage: string;
}

export interface AppIdentity {
  app: string;
}

export interface Identity extends StackStageIdentity, AppIdentity {}

export const AppIdentity = {
  suffixText(appIdentity: AppIdentity, text: string): string {
    const titleCaseApp = appIdentity.app.charAt(0).toUpperCase() + appIdentity.app.slice(1);
    return `${text}${titleCaseApp}`;
  },
  taggedConstruct<T extends IConstruct>(appIdentity: AppIdentity, construct: T): T {
    Tags.of(construct).add("App", appIdentity.app);
    return construct;
  },
};
