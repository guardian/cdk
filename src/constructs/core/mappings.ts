import { CfnMapping } from "@aws-cdk/core";
import type { Stage } from "../../constants";
import type { AppIdentity } from "./identity";
import type { GuStack } from "./stack";

export type GuMappingValue = string | number | boolean;

export interface GuStageMappingValue<T extends GuMappingValue> extends AppIdentity {
  variableName: string;
  stageValues: Record<Stage, T>;
}

export class GuStageMapping {
  private readonly scope: GuStack;

  // the key in this Map is an app string
  private mappings: Map<string, CfnMapping>;

  constructor(scope: GuStack) {
    this.scope = scope;
    this.mappings = new Map<string, CfnMapping>();
  }

  private getCfnMapping(app: string): CfnMapping {
    if (!this.mappings.has(app)) {
      const emptyCfnMapping = new CfnMapping(this.scope, app);
      this.mappings.set(app, emptyCfnMapping);
      return emptyCfnMapping;
    }

    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion -- `this.mappings.has` above is `true`, so we can safely assert non-null
    return this.mappings.get(app)!;
  }

  withValue<T extends GuMappingValue>(mappingValue: GuStageMappingValue<T>): T {
    const { app, variableName, stageValues } = mappingValue;

    const cfnMapping = this.getCfnMapping(app);

    for (const [stage, value] of Object.entries(stageValues)) {
      cfnMapping.setValue(stage, variableName, value);
    }

    return cfnMapping.findInMap(this.scope.stage, variableName) as T;
  }
}
