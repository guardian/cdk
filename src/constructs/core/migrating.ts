import { Annotations } from "@aws-cdk/core";
import { isGuStatefulConstruct } from "../../utils/mixin";
import type { CfnElement, IConstruct } from "@aws-cdk/core";

export interface GuMigratingStack {
  /**
   * A flag to symbolise if a stack is being migrated from a previous format (eg YAML) into guardian/cdk.
   * A value of `true` means resources in the stack can have custom logicalIds set using the property `existingLogicalId` (where available).
   * A value of `false` or `undefined` means the stack is brand new. Any resource that gets created will have an auto-generated logicalId.
   * Ideally, for use only by [[ `GuStack` ]].
   * @see GuMigratingResource
   * @see GuStack
   */
  migratedFromCloudFormation: boolean;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/explicit-module-boundary-types -- user defined type guard
export function isGuMigratingStack(construct: any): construct is GuMigratingStack {
  return "migratedFromCloudFormation" in construct;
}

export interface GuMigratingResource {
  /**
   * A string to use to override the logicalId AWS CDK auto-generates for a resource.
   * This is useful when migrating a pre-existing stack into guardian/cdk,
   * as it ensures resources are kept rather than recreated.
   * For example, imagine a YAML stack that creates a load balancer with logicalId `DotcomLoadbalancer`.
   * We would want to set `existingLogicalId` for the GuLoadBalancer in guardian/cdk to ensure it is preserved when moving to guardian/cdk.
   * Otherwise it will be created as something like `DotcomLoadbalancerABCDEF`,
   * is a new resource, and require any DNS entries to be updated accordingly.
   *
   * @see GuMigratingStack
   * @see GuStack
   */
  existingLogicalId?: string;
}

export const GuMigratingResource = {
  setLogicalId<T extends IConstruct>(
    construct: T,
    { migratedFromCloudFormation }: GuMigratingStack,
    { existingLogicalId }: GuMigratingResource
  ): void {
    const overrideLogicalId = (logicalId: string) => {
      const defaultChild = construct.node.defaultChild as CfnElement;
      defaultChild.overrideLogicalId(logicalId);
    };

    const id = construct.node.id;
    const isStateful = isGuStatefulConstruct(construct);

    if (migratedFromCloudFormation) {
      if (existingLogicalId) {
        return overrideLogicalId(existingLogicalId);
      }

      if (isStateful) {
        Annotations.of(construct).addWarning(
          `GuStack has 'migratedFromCloudFormation' set to true. ${id} is a stateful construct and 'existingLogicalId' has not been set. ${id}'s logicalId will be auto-generated and consequently AWS will create a new resource rather than inheriting an existing one. This is not advised as downstream services, such as DNS, will likely need updating.`
        );
      }
    } else {
      if (existingLogicalId) {
        Annotations.of(construct).addWarning(
          `GuStack has 'migratedFromCloudFormation' set to false. ${id} has an 'existingLogicalId' set to ${existingLogicalId}. This will have no effect - the logicalId will be auto-generated. Set 'migratedFromCloudFormation' to true for 'existingLogicalId' to be observed.`
        );
      }

      if (isStateful) {
        Annotations.of(construct).addInfo(
          `GuStack has 'migratedFromCloudFormation' set to false. ${id} is a stateful construct, it's logicalId will be auto-generated and AWS will create a new resource.`
        );
      }
    }
  },
};
