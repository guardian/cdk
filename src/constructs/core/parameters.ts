import type { CfnParameterProps } from "@aws-cdk/core";
import { CfnParameter } from "@aws-cdk/core";
import { RegexPattern, Stage, Stages } from "../../constants";
import type { GuStack } from "./stack";

export type GuParameterProps = CfnParameterProps;

export type GuNoTypeParameterProps = Omit<GuParameterProps, "type">;

export class GuParameter extends CfnParameter {
  constructor(scope: GuStack, id: string, props: GuParameterProps) {
    super(scope, id, props);
  }
}

export class GuStringParameter extends GuParameter {
  constructor(scope: GuStack, id: string, props: GuNoTypeParameterProps) {
    super(scope, id, { ...props, type: "String" });
  }
}

export class GuStageParameter extends GuParameter {
  constructor(scope: GuStack, id: string = "Stage") {
    super(scope, id, {
      type: "String",
      description: "Stage name",
      allowedValues: Stages,
      default: Stage.CODE,
    });
  }
}

export class GuStackParameter extends GuParameter {
  constructor(scope: GuStack, id: string = "Stack") {
    super(scope, id, {
      type: "String",
      description: "Name of this stack",
      default: "deploy",
    });
  }
}

export class GuInstanceTypeParameter extends GuParameter {
  constructor(scope: GuStack, id: string = "InstanceType", props: GuParameterProps = {}) {
    super(scope, id, {
      type: "String",
      description: "EC2 Instance Type",
      default: "t3.small",
      ...props,
    });
  }
}

export class GuSSMParameter extends GuParameter {
  constructor(scope: GuStack, id: string, props: GuNoTypeParameterProps) {
    super(scope, id, {
      ...props,
      type: "AWS::SSM::Parameter::Value<String>",
    });
  }
}

export class GuSubnetListParameter extends GuParameter {
  constructor(scope: GuStack, id: string, props: GuNoTypeParameterProps) {
    super(scope, id, { ...props, type: "List<AWS::EC2::Subnet::Id>" });
  }
}

export class GuVpcParameter extends GuParameter {
  constructor(scope: GuStack, id: string, props: GuNoTypeParameterProps) {
    super(scope, id, {
      ...props,
      type: "AWS::EC2::VPC::Id",
    });
  }
}

export class GuAmiParameter extends GuParameter {
  constructor(scope: GuStack, id: string, props: GuNoTypeParameterProps) {
    super(scope, id, {
      ...props,
      type: "AWS::EC2::Image::Id",
    });
  }
}

export class GuArnParameter extends GuStringParameter {
  constructor(scope: GuStack, id: string, props: GuNoTypeParameterProps) {
    super(scope, id, {
      ...props,
      allowedPattern: RegexPattern.ARN,
      constraintDescription: "Must be a valid ARN, eg: arn:partition:service:region:account-id:resource-id",
    });
  }
}

export class GuS3ObjectArnParameter extends GuStringParameter {
  constructor(scope: GuStack, id: string, props: GuNoTypeParameterProps) {
    super(scope, id, {
      ...props,
      allowedPattern: RegexPattern.S3ARN,
      constraintDescription:
        "Must be a valid S3 ARN, see https://docs.aws.amazon.com/AmazonS3/latest/dev/BucketRestrictions.html",
    });
  }
}

export class GuGuardianEmailSenderParameter extends GuStringParameter {
  constructor(scope: GuStack, id: string = "EmailSenderAddress", props?: GuNoTypeParameterProps) {
    super(scope, id, {
      ...props,
      allowedPattern: RegexPattern.GUARDIAN_EMAIL,
      constraintDescription: "Must be an @theguardian.com email address",
    });
  }
}
