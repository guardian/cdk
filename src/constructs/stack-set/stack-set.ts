import { CfnStackSet } from "@aws-cdk/core";
import type { GuStack, GuStackForStackSetInstance } from "../core";

interface GuStackSetProps {
  /**
   * The contents for `TemplateBody` within a `AWS::CloudFormation::StackSet`.
   *
   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-cloudformation-stackset.html#cfn-cloudformation-stackset-templatebody
   */
  stackSetInstance: GuStackForStackSetInstance;

  /**
   * The name of the stack set. This will appear in the target account.
   */
  name: string;

  /**
   * A description of the stack set. This will appear in the target account.
   */
  description: string;

  /**
   * AWS Organisation Unit IDs to deploy the stack set into.
   *
   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-cloudformation-stackset-deploymenttargets.html#cfn-cloudformation-stackset-deploymenttargets-organizationalunitids
   */
  organisationUnitTargets: string[];

  /**
   * The contents for`Parameters` within a `AWS::CloudFormation::StackSet`.
   *
   * Typically, you'll only need parameters if the parent stack creates resources for use in the stack set instances.
   *
   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-cloudformation-stackset.html#cfn-cloudformation-stackset-parameters
   */
  stackSetInstanceParameters?: Record<string, string>;

  /**
   * Which regions to run the stack set in.
   *
   * @default The region of the parent stack.
   */
  regions?: string[];
}

/**
 * A construct to create a `AWS::CloudFormation::StackSet`.
 *
 * The stack set will be configured to automatically deploy into accounts when they join an AWS Organisation Unit.
 * It's assumed stack sets are only provisioned for infrastructure, therefore the `STAGE` value will always be `INFRA`.
 *
 * Usage:
 * ```typescript
 * // the infrastructure to create in target accounts
 * class AccountAlertTopic extends GuStackForStackSetInstance {
 *   constructor(id: string, props: GuStackProps) {
 *     super(id, props);
 *
 *     new GuSnsTopic(this, "topic-for-alerts");
 *   }
 * }
 *
 * class ParentStack extends GuStackForInfrastructure {
 *   constructor(scope: App, id: string, props: GuStackProps) {
 *     super(scope, id, props);
 *
 *     new GuStackSet(this, `${id}StackSet`, {
 *       stackSetInstance: new AccountAlertTopic("Alerts", { stack: props.stack }),
 *       name: "centralised-alarms",
 *       description: "Provisioning of standard account alerting resources",
 *       organisationUnitTargets: [ "o-abcde12345" ]
 *     });
 *   }
 * }
 *
 * // contents of `bin/cdk.ts`
 * new ParentStack(new App(), "AccountAlarmResources", { stack: "alarms" })
 * ```
 *
 * This will produce a CloudFormation template like this:
 * ```yaml
 * Resources:
 *   AccountAlarmResourcesStackSet:
 *     Type: AWS::CloudFormation::StackSet
 *     Properties:
 *       PermissionModel: SERVICE_MANAGED
 *       StackSetName: centralised-alarms
 *       AutoDeployment:
 *         Enabled: true
 *         RetainStacksOnAccountRemoval: false
 *       Description: Provisioning of standard account alerting resources
 *       Parameters: []
 *       StackInstancesGroup:
 *         - DeploymentTargets:
 *             OrganizationalUnitIds:
 *               - o-abcde12345
 *           Regions:
 *             - Ref: AWS::Region
 *       Tags:
 *         - Key: gu:cdk:version
 *           Value: TEST
 *         - Key: gu:repo
 *           Value: guardian/cdk
 *         - Key: Stack
 *           Value: alarms
 *         - Key: Stage
 *           Value: INFRA
 *       TemplateBody: |-
 *         {
 *           "Resources": {
 *             "topicforalerts57330FBE": {
 *               "Type": "AWS::SNS::Topic",
 *               "Properties": {
 *                 "Tags": [
 *                   {
 *                     "Key": "gu:cdk:version",
 *                     "Value": "TEST"
 *                   },
 *                   {
 *                     "Key": "gu:repo",
 *                     "Value": "guardian/cdk"
 *                   },
 *                   {
 *                     "Key": "Stack",
 *                     "Value": "alarms"
 *                   },
 *                   {
 *                     "Key": "Stage",
 *                     "Value": "INFRA"
 *                   }
 *                 ]
 *               }
 *             }
 *           }
 *         }
 * ```
 *
 * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-cloudformation-stackset.html
 */
export class GuStackSet extends CfnStackSet {
  constructor(scope: GuStack, id: string, props: GuStackSetProps) {
    const stackSetInstanceParameters = props.stackSetInstanceParameters ?? {};

    const params = Object.keys(stackSetInstanceParameters);
    const undefinedStackSetParams = props.stackSetInstance.parameterKeys.filter((_) => !params.includes(_));

    if (undefinedStackSetParams.length !== 0) {
      throw new Error(`There are undefined stack set parameters: ${undefinedStackSetParams.join(", ")}`);
    }

    super(scope, id, {
      stackSetName: props.name,
      description: props.description,
      permissionModel: "SERVICE_MANAGED",
      autoDeployment: {
        enabled: true,
        retainStacksOnAccountRemoval: false,
      },
      stackInstancesGroup: [
        {
          regions: props.regions ?? [scope.region],
          deploymentTargets: {
            organizationalUnitIds: props.organisationUnitTargets,
          },
        },
      ],
      templateBody: props.stackSetInstance.cfnJson,
      parameters: Object.entries(stackSetInstanceParameters).map(([key, value]) => {
        return { parameterKey: key, parameterValue: value };
      }),
    });
  }
}
