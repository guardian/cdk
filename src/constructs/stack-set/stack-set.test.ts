import { SynthUtils } from "@aws-cdk/assert";
import { OrganizationPrincipal } from "@aws-cdk/aws-iam";
import { Topic } from "@aws-cdk/aws-sns";
import { RegexPattern } from "../../constants";
import { simpleGuStackForTesting } from "../../utils/test";
import { GuStackForStackSetInstance, GuStringParameter } from "../core";
import { GuKinesisStream } from "../kinesis";
import { GuSnsTopic } from "../sns";
import { GuStackSet } from "./stack-set";

describe("The GuStackSet construct", () => {
  it("should correctly provision a stack with a stack set resource", () => {
    const theStackSetInstance = new GuStackForStackSetInstance("the-stack-set", { stack: "test", stage: "TEST" });
    new GuKinesisStream(theStackSetInstance, "account-logging-stream");

    const parentStack = simpleGuStackForTesting({ stack: "test" });

    new GuStackSet(parentStack, "StackSet", {
      name: "my-stack-set",
      description: "this stack set provisions some common infrastructure",
      organisationUnitTargets: ["o-12345abcde"],
      stackSetInstance: theStackSetInstance,
    });

    expect(SynthUtils.toCloudFormation(parentStack)).toMatchSnapshot();
  });

  it("should support parameters in the stack set instance", () => {
    const theStackSetInstance = new GuStackForStackSetInstance("the-stack-set", { stack: "test", stage: "TEST" });
    Topic.fromTopicArn(
      theStackSetInstance,
      "central-topic",
      new GuStringParameter(theStackSetInstance, "CentralSnsTopicArn", { allowedPattern: RegexPattern.ARN })
        .valueAsString
    );

    const awsOrgId = "o-12345abcde";
    const parentStack = simpleGuStackForTesting({ stack: "test" });
    const centralTopic = new GuSnsTopic(parentStack, "account-alerts");
    centralTopic.grantPublish(new OrganizationPrincipal(awsOrgId));

    new GuStackSet(parentStack, "StackSet", {
      name: "my-stack-set",
      description: "this stack set provisions some common infrastructure",
      organisationUnitTargets: [awsOrgId],
      stackSetInstance: theStackSetInstance,
      stackSetInstanceParameters: {
        CentralSnsTopicArn: centralTopic.topicArn,
      },
    });

    expect(SynthUtils.toCloudFormation(parentStack)).toMatchSnapshot();
  });

  it("should error if the parent stack does not specify all parameters for the stack set instance template", () => {
    const theStackSetInstance = new GuStackForStackSetInstance("the-stack-set", { stack: "test", stage: "TEST" });
    new GuStringParameter(theStackSetInstance, "CentralSnsTopic", { allowedPattern: RegexPattern.ARN });

    const parentStack = simpleGuStackForTesting({ stack: "test" });

    expect(() => {
      new GuStackSet(parentStack, "StackSet", {
        name: "my-stack-set",
        description: "this stack set provisions some common infrastructure",
        organisationUnitTargets: ["o-12345abcde"],
        stackSetInstance: theStackSetInstance,
      });
    }).toThrow("There are undefined stack set parameters: CentralSnsTopic");
  });
});
