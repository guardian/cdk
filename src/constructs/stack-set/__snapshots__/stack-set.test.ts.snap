// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The GuStackSet construct should correctly provision a stack with a stack set resource 1`] = `
Object {
  "Resources": Object {
    "StackSet": Object {
      "Properties": Object {
        "AutoDeployment": Object {
          "Enabled": true,
          "RetainStacksOnAccountRemoval": false,
        },
        "Description": "this stack set provisions some common infrastructure",
        "Parameters": Array [],
        "PermissionModel": "SERVICE_MANAGED",
        "StackInstancesGroup": Array [
          Object {
            "DeploymentTargets": Object {
              "OrganizationalUnitIds": Array [
                "o-12345abcde",
              ],
            },
            "Regions": Array [
              Object {
                "Ref": "AWS::Region",
              },
            ],
          },
        ],
        "StackSetName": "my-stack-set",
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          Object {
            "Key": "Stack",
            "Value": "test",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "TemplateBody": "{
  \\"Resources\\": {
    \\"accountloggingstreamB8733874\\": {
      \\"Type\\": \\"AWS::Kinesis::Stream\\",
      \\"Properties\\": {
        \\"RetentionPeriodHours\\": 24,
        \\"ShardCount\\": 1,
        \\"StreamEncryption\\": {
          \\"Fn::If\\": [
            \\"AwsCdkKinesisEncryptedStreamsUnsupportedRegions\\",
            {
              \\"Ref\\": \\"AWS::NoValue\\"
            },
            {
              \\"EncryptionType\\": \\"KMS\\",
              \\"KeyId\\": \\"alias/aws/kinesis\\"
            }
          ]
        },
        \\"StreamModeDetails\\": {
          \\"StreamMode\\": \\"PROVISIONED\\"
        },
        \\"Tags\\": [
          {
            \\"Key\\": \\"gu:cdk:version\\",
            \\"Value\\": \\"TEST\\"
          },
          {
            \\"Key\\": \\"gu:repo\\",
            \\"Value\\": \\"guardian/cdk\\"
          },
          {
            \\"Key\\": \\"Stack\\",
            \\"Value\\": \\"test\\"
          },
          {
            \\"Key\\": \\"Stage\\",
            \\"Value\\": \\"TEST\\"
          }
        ]
      }
    }
  },
  \\"Conditions\\": {
    \\"AwsCdkKinesisEncryptedStreamsUnsupportedRegions\\": {
      \\"Fn::Or\\": [
        {
          \\"Fn::Equals\\": [
            {
              \\"Ref\\": \\"AWS::Region\\"
            },
            \\"cn-north-1\\"
          ]
        },
        {
          \\"Fn::Equals\\": [
            {
              \\"Ref\\": \\"AWS::Region\\"
            },
            \\"cn-northwest-1\\"
          ]
        }
      ]
    }
  }
}",
      },
      "Type": "AWS::CloudFormation::StackSet",
    },
  },
}
`;

exports[`The GuStackSet construct should support parameters in the stack set instance 1`] = `
Object {
  "Resources": Object {
    "StackSet": Object {
      "Properties": Object {
        "AutoDeployment": Object {
          "Enabled": true,
          "RetainStacksOnAccountRemoval": false,
        },
        "Description": "this stack set provisions some common infrastructure",
        "Parameters": Array [
          Object {
            "ParameterKey": "CentralSnsTopicArn",
            "ParameterValue": Object {
              "Ref": "accountalertsD9982E6F",
            },
          },
        ],
        "PermissionModel": "SERVICE_MANAGED",
        "StackInstancesGroup": Array [
          Object {
            "DeploymentTargets": Object {
              "OrganizationalUnitIds": Array [
                "o-12345abcde",
              ],
            },
            "Regions": Array [
              Object {
                "Ref": "AWS::Region",
              },
            ],
          },
        ],
        "StackSetName": "my-stack-set",
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          Object {
            "Key": "Stack",
            "Value": "test",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "TemplateBody": "{
  \\"Parameters\\": {
    \\"CentralSnsTopicArn\\": {
      \\"Type\\": \\"String\\",
      \\"AllowedPattern\\": \\"arn:aws:[a-z0-9]*:[a-z0-9\\\\\\\\-]*:[0-9]{12}:.*\\"
    }
  }
}",
      },
      "Type": "AWS::CloudFormation::StackSet",
    },
    "accountalertsD9982E6F": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          Object {
            "Key": "Stack",
            "Value": "test",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::SNS::Topic",
    },
    "accountalertsPolicy8E37A8FA": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sns:Publish",
              "Condition": Object {
                "StringEquals": Object {
                  "aws:PrincipalOrgID": "o-12345abcde",
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "AWS": "*",
              },
              "Resource": Object {
                "Ref": "accountalertsD9982E6F",
              },
              "Sid": "0",
            },
          ],
          "Version": "2012-10-17",
        },
        "Topics": Array [
          Object {
            "Ref": "accountalertsD9982E6F",
          },
        ],
      },
      "Type": "AWS::SNS::TopicPolicy",
    },
  },
}
`;
